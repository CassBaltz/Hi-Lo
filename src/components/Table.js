import React from "react";
import { withStyles } from "@material-ui//styles";
import { connect } from "react-redux";
import { guess, passTurn } from '../redux/actions/game';
import { LOWER, HIGHER, CORRECT_GUESSES_TIL_PASS } from '../redux/constants';

import DiscardPile from "./DiscardPile";

const CurrentGuess = ({ classes, card, }) => {
  if (card) {
    return (
      <img
        className={classes.currentGuessCard}
        src={card.image}
        alt={card.code}
      />
    );
  }

  return (
    <div className={classes.currentGuessPlaceholder}></div>
  );
}

export const Table = ({
  classes,
  dispatch,
  cardsShowing,
  correctGuesses,
  cardsRemainingInDeck,
  currentGuess,
  resettingDeck,
}) => {
  const canPassTurn = correctGuesses >= CORRECT_GUESSES_TIL_PASS;
  const { guessedCard } = currentGuess;
  // disable all actions when we are in the
  // 'guessing' state
  const enableButtons = !guessedCard;

  return (
    <div className={classes.root}>
      <button
        className={classes.higherButton}
        onClick={() => enableButtons && dispatch(guess(HIGHER))}
      >
        HIGHER
      </button>
      <div className={classes.inlineItems}>
        <div className={classes.cardDeck}>{!resettingDeck && cardsRemainingInDeck}</div>
        <CurrentGuess card={guessedCard} classes={classes} />
        <DiscardPile cardsShowing={cardsShowing} />
        <button
          className={classes.passTurnButton}
          disabled={!canPassTurn}
          onClick={() => enableButtons && canPassTurn && dispatch(passTurn())}
        >
          PASS TURN
        </button>
      </div>
      <button
        className={classes.lowerButton}
        onClick={() => enableButtons && dispatch(guess(LOWER))}
      >
        LOWER
      </button>
    </div>
  );
};

const colorButtonFromProps = (props, buttonDirection) => {
  const { isCorrect, direction } = props.currentGuess;
  if (direction === buttonDirection) {
    return isCorrect ? "#63e6be" : "#ffa8a8";
  }

  return sharedButtonStyles.backgroundColor;
};

const sharedButtonStyles = {
  display: "flex",
  width: "70px",
  padding: "10px",
  backgroundColor: "#c5f6fa",
  color: "#343a40",
  position: "absolute",
  boxShadow: "unset",
  border: "unset",
  justifyContent: "center",
  "&:focus": {
    outline: "none"
  },
  "&:hover": {
    backgroundColor: "#e3fafc"
  },
  cursor: "pointer",
  transition: "background-color .5s",
  borderRadius: "4px",
};

const currentGuessShared = {
  width: "70px",
  height: "98px",
  position: "relative",
  left: "60px",
};

const styles = {
  root: {
    backgroundColor: "#fff4e6",
    border: "8px solid #fd7e14",
    width: "500px",
    height: "325px",
    borderRadius: "20px",
    margin: "10px auto",
    position: "relative",
    display: "flex",
    alignItems: "center"
  },
  inlineItems: {
    position: "relative",
    margin: "auto",
    display: "flex",
    alignItems: "center",
    width: "100%",
  },
  cardDeck: {
    width: "70px",
    height: "98px",
    backgroundColor: props => props.resettingDeck ? "#f1f3f5" : "#ff6b6b",
    borderRadius: "4px",
    border: "2px solid #FFF",
    boxSizing: "border-box",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    color: "#FFF",
    fontWeight: "700",
    fontSize: "30px",
    position: "relative",
    left: "20px"
  },
  higherButton: {
    ...sharedButtonStyles,
    left: "20px",
    top: "20px",
    backgroundColor: props => colorButtonFromProps(props, HIGHER),
    "&:hover": {
      backgroundColor: props => colorButtonFromProps(props, HIGHER),
    },
  },
  lowerButton: {
    ...sharedButtonStyles,
    left: "20px",
    bottom: "20px",
    backgroundColor: props => colorButtonFromProps(props, LOWER),
    "&:hover": {
      backgroundColor: props => colorButtonFromProps(props, LOWER),
    },
  },
  passTurnButton: {
    ...sharedButtonStyles,
    "&:disabled": {
      backgroundColor: "#e3fafc",
      opacity: ".5 !important",
      cursor: "not-allowed",
    },
    position: "relative",
    left: "120px",
  },
  currentGuessCard: {
    ...currentGuessShared,
  },
  currentGuessPlaceholder: {
    ...currentGuessShared,
    background: "transparent",
    boxSizing: "border-box",
    border: "1px solid #ced4da",
    borderRadius: "3px",
  },
};

export const mapStateToProps = ({ game, deck }) => ({
  resettingDeck: deck.resettingDeck,
  cardsShowing: deck.cardsShowing,
  cardsRemainingInDeck: deck.cards.length,
  correctGuesses: game.turn.correctGuesses,
  currentGuess: game.currentGuess,

});

export default connect(mapStateToProps, null)(
  withStyles(styles)(Table)
);


