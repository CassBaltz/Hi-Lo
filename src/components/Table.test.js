import React from "react";
import { Table, styles } from "./Table";

const props = {
  classes: classReducer(styles),
  dispatch: jest.fn(),
  cardsShowing: 1,
  correctGuesses: 0,
  cardsRemainingInDeck: 10,
  currentGuess: {
    direction: null,
    isCorrect: false,
    guessedCard: null,
  },
  resettingDeck: false,
};

describe("<Table />", () => {
  it("should render correctly with the default props", () => {
    const table = shallow(<Table {...props} />);
    expect(toJson(table)).toMatchSnapshot();
  });

  it("should prevent guesses when there is a guessed card", () => {
    const guessedCardProps = {
      ...props,
      currentGuess: {
        ...props.currentGuess,
        guessedCard: {
          code: "AD",
          image: "ace image",
          value: "ACE"
        }
      }
    };
    const table = shallow(<Table {...guessedCardProps} />);
    const higherButton = table.find("button.higherButton");
    higherButton.simulate("click");

    expect(props.dispatch).not.toHaveBeenCalled();
  });

  it("should enable the 'PASS TURN' button when there are three correct guesses", () => {
    const table = shallow(<Table {...props} />);
    expect(table.find("button.passTurnButton").prop("disabled")).toBe(true);

    table.setProps({ correctGuesses: 3 });
    expect(table.find("button.passTurnButton").prop("disabled")).toBe(false);
  });
});