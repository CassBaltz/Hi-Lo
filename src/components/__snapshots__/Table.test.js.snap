// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render correctly with the default props 1`] = `
<div
  className="root"
>
  <button
    className="higherButton"
    onClick={[Function]}
  >
    HIGHER
  </button>
  <div
    className="inlineItems"
  >
    <div
      className="cardDeck"
    >
      10
    </div>
    <CurrentGuess
      card={null}
      classes={
        Object {
          "cardDeck": "cardDeck",
          "currentGuessCard": "currentGuessCard",
          "currentGuessPlaceholder": "currentGuessPlaceholder",
          "higherButton": "higherButton",
          "inlineItems": "inlineItems",
          "lowerButton": "lowerButton",
          "passTurnButton": "passTurnButton",
          "root": "root",
        }
      }
    />
    <WithStyles(DiscardPile)
      cardsShowing={1}
    />
    <button
      className="passTurnButton"
      disabled={true}
      onClick={[Function]}
    >
      PASS TURN
    </button>
  </div>
  <button
    className="lowerButton"
    onClick={[Function]}
  >
    LOWER
  </button>
</div>
`;
